Bluetooth: controller: Do not use LL procedures not supported by remote

joerchan opened the pull request:
Update LLCP handling during PHY update and Data Length update to not
start the LL control procedure if the remote has already indicated that
the procedure is not supported.

This fulfills the following requirement from the BT Core Specification
(Core_v5.2, Vol 6, Part B, Section 4.6):

Except where explicitly stated elsewhere in this specification, if the
peer Link Layer has indicated either during a feature exchange procedure
or by responding with an LL_UNKNOWN_RSP PDU that it does not support a
procedure, then the Link Layer shall not use that procedure.

Re-use the connection parameter request handling for PHY and
data length update procedures.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1268,7 +1268,9 @@ static void conn_auto_initiate(struct bt_conn *conn)
-	    BT_FEAT_LE_PHY_2M(bt_dev.le.features)) {
+	    BT_FEAT_LE_PHY_2M(bt_dev.le.features) &&
+	    !(atomic_test_bit(conn->flags, BT_CONN_AUTO_FEATURE_EXCH) &&
+	      !BT_FEAT_LE_PHY_2M(conn->le.features))) {
```
I'm having a bit of a hard time wrapping my head around this logic. Would the following make any more sense:
```
        (!atomic_test_bit(conn->flags, BT_CONN_AUTO_FEATURE_EXCH) ||
         BT_FEAT_LE_PHY_2M(conn->le.features))) {
```

I.e. remove the outer negation, negate the internal terms and change the logical AND to a logical OR.

joerchan replied:
Yeah, it looks better. Updated.

rugeGerritsen commented:
The BLE host does not need this change, as it should be taken care of by the the controller. See Core_v5.2, Vol 6, Part B, Section 4.6:
_Except where explicitly stated elsewhere in this specification, if the peer Link
Layer has indicated either during a feature exchange procedure or by
responding with an LL_UNKNOWN_RSP PDU that it does not support a
procedure, then the Link Layer shall not use that procedure._

joerchan commented:
> The BLE host does not need this change, as it should be taken care of by the the controller. See Core_v5.2, Vol 6, Part B, Section 4.6:
> _Except where explicitly stated elsewhere in this specification, if the peer Link Layer has indicated either during a feature exchange procedure or by responding with an LL_UNKNOWN_RSP PDU that it does not support a procedure, then the Link Layer shall not use that procedure._

@rugeGerritsen
So you are saying that the Controller should not have started the procedure at all? Because that is not what is happening. Should we report this as a LL bug then ?
Also will there be a phy_changed/data_length_changed event in this case? Because the host handling here relies on there being one. Could you clarify this?
If there is no event then probably the best thing to do would be to initiate all the procedures immediately, but I'm not sure the Zephyr LL supports this.

rugeGerritsen commented:
> The BLE host does not need this change, as it should be taken care of by the the controller. See Core_v5.2, Vol 6, Part B, Section 4.6:
> Except where explicitly stated elsewhere in this specification, if the peer Link Layer has indicated either during a feature exchange procedure or by responding with an LL_UNKNOWN_RSP PDU that it does not support a procedure, then the Link Layer shall not use that procedure.
> 
> @rugeGerritsen
> So you are saying that the Controller should not have started the procedure at all? Because that is not what is happening. Should we report this as a LL bug then ?
> Also will there be a phy_changed/data_length_changed event in this case? Because the host handling here relies on there being one. Could you clarify this?
> If there is no event then probably the best thing to do would be to initiate all the procedures immediately, but I'm not sure the Zephyr LL supports this.

@joerchan, you are right. This is a controller bug. From _LL/CON/SLA/BV-85-C, Initiating Connection Parameter Request - Unsupported Without Feature Exchange_ it is indicated that there are two accepted behaviors. Either return 0x1A upon initiation of the control procedure, or return the event with an error code.

joerchan commented:
Thank you for the clarification. You meant LL/CON/SLA/BV-86-C [Initiating Connection Parameter Request – Unsupported With
Feature Exchange]. 
That is for connection parameters request, I couldn't find a matching test for data length or phy procedure, but I think it is fair to assume that the same would be expected in this case.

cvinayak commented:
During the development of DLE and PHY update procedure, I have noticed phones in the market that support these procedures without the feature bits set.

Wrt to CPR, its covered in the controller, by falling back to using connection update procedure:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ull_conn.c#L5783
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ull_conn.c#L276

joerchan commented:
@cvinayak But is it still the case? That sounds like testing before qualification.

joerchan commented:
In my opinion the right approach would be to add the same handling that CPR has to data length and phy update procedures in the controller.
Surely by now the devices that support these features have set the bits.
@cvinayak @jhedberg Opinions on this?

Also vinayak, is there still a need to wait for the previous procedure to complete before starting the next one?
@rugeGerritsen Same question to you really. Would it be possible to just send all commands immediately to the controller?

rugeGerritsen commented:
> In my opinion the right approach would be to add the same handling that CPR has to data length and phy update procedures in the controller.
> Surely by now the devices that support these features have set the bits.
> @cvinayak @jhedberg Opinions on this?
> Also vinayak, is there still a need to wait for the previous procedure to complete before starting the next one?
> @rugeGerritsen Same question to you really. Would it be possible to just send all commands immediately to the controller?

@joerchan , For the Nordic Controller it is possible to initiate all control procedures at once, and I assume that the host should also assume that this is the case.

joerchan committed:
Bluetooth: controller: Do not use LL procedures not supported by remote

Update LLCP handling during PHY update and Data Length update to not
start the LL control procedure if the remote has already indicated that
the procedure is not supported.

This fulfills the following requirement from the BT Core Specification
(Core_v5.2, Vol 6, Part B, Section 4.6):

Except where explicitly stated elsewhere in this specification, if the
peer Link Layer has indicated either during a feature exchange procedure
or by responding with an LL_UNKNOWN_RSP PDU that it does not support a
procedure, then the Link Layer shall not use that procedure.

Re-use the connection parameter request handling for PHY and
data length update procedures.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -684,6 +685,7 @@ u8_t ll_adv_enable(u8_t enable)
 
 #if defined(CONFIG_BT_CTLR_PHY)
 		conn->llcp_phy.req = conn->llcp_phy.ack = 0;
+		conn->llcp_phy.disabled = 0;
```
Missing U?

joerchan replied:
Updated.

Note: there are other cases where U is missing. You might want to got over it.
I really think this part of the code could benefit from a memset to zero.

cvinayak commented:
> Also vinayak, is there still a need to wait for the previous procedure to complete before starting the next one?

Zephyr controller will not handle this correctly, this is something postponed to the new LLCP work.

joerchan commented:
> > Also vinayak, is there still a need to wait for the previous procedure to complete before starting the next one?
> 
> Zephyr controller will not handle this correctly, this is something postponed to the new LLCP work.

Will leave it as is then.

cvinayak commented:
> Will leave it as is then.

@joerchan It is not a big change to achieve queuing (change of if-then-else to switch-case on a started local control procedure). New LLCP will take until v2.4 to mature, if I get motivated in the v2.3 timeframe, I will get back to you if I have anything.

jhedberg commented:
@joerchan could you update the PR title and description to match the content? It's confusing to see "host" in the title even though the content is now purely controller related. Btw, is this something we want for 2.2? (there's no referenced issue and the milestone isn't set, so it's hard for me to say)

joerchan commented:
@jhedberg This is a bu, but only a technicality, can be left out of 2.2
